### PR TITLE
fix(connector): dbz source handler avoid oom more elegantly

### PR DESCRIPTION
**Pull request by WillyKidd**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## Background
- previously, we observed that out of memory (oom) error will occur if NO throttling is implemented in the debezium source handler (the server of the rpc/the producer/the sender)
- if the client of the rpc consumes messages slower than the server produces, the sender's buffer will get larger and larger (back pressure) if no throttling is implemented, which eventually results in oom in the sender
- grpc-java provides a method `isReady`, which returns `true` whenever the sender's buffer is less than 32kB (on ready threshold)
- the on ready threshold has been fixed to 32kB in grpc-java (written as `static final`)

References:
- [On Ready Threshold fixed to 32kB](https://github.com/grpc/grpc-java/issues/5433)
- [onReadyHandler example from grpc-java contributor](https://github.com/grpc/grpc-java/issues/2247#issuecomment-245949881)
- [Some helpful explanations of buffering in grpc](https://github.com/salesforce/reactive-grpc/issues/255)

## What's changed and what's your intention?

- previously, we used `sleep` to wait for the sender's buffer size to be smaller than 32 kB (`isReady` becomes `true`), and it turned out to hurt the throughput of the connector (in our test, the throughput drops to around 2,000 rows/second, whereas the throughput is around 25,000 rows/second without throttling)
- therefore in this pr, we switch to `OnReadyHandler`, which avoids oom and provides better performance
- `OnReadyHandler` is called whenever the value of `isReady` changes from `false` to `true`

## Checklist For Contributors

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
~~- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).~~
~~- [ ] I have demonstrated that backward compatibility is not broken by breaking changes and created issues to track deprecated features to be removed in the future. (Please refer to the issue)~~
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Checklist For Reviewers

- [x] I have requested macro/micro-benchmarks as this PR can affect performance substantially, and the results are shown.

## Documentation

- [x] My PR **DOES NOT** contain user-facing changes.

<details><summary>Click here for Documentation</summary>

### Types of user-facing changes

Please keep the types that apply to your changes, and remove the others.

- Installation and deployment
- Connector (sources & sinks)
- SQL commands, functions, and operators
- RisingWave cluster configuration changes
- Other (please specify in the release note below)

### Release note

</details>

### CODE REVIEW

File: java/connector-node/risingwave-connector-service/src/main/java/com/risingwave/sourcenode/core/DbzSourceHandler.java [hunk 1]
```diff
@@ -46,39 +46,57 @@ public void startSource(ServerCallStreamObserver<GetEventStreamResponse> respons
...
             runner.start();
             LOG.info("Start consuming events of table {}", config.getSourceId());
-            while (runner.isRunning()) {
-                try {
-                    // Thread will block on the channel to get output from engine
-                    var resp =
-                            runner.getEngine().getOutputChannel().poll(500, TimeUnit.MILLISECONDS);
 
-                    if (resp != null) {
+            class OnReadyHandler implements Runnable {
```
Comment (StrikeW) [inline java/connector-node/risingwave-connector-service/src/main/java/com/risingwave/sourcenode/core/DbzSourceHandler.java:50]: May move the class definition out of the function to make the code easy to read.

File: java/connector-node/risingwave-connector-service/src/main/java/com/risingwave/sourcenode/core/DbzSourceHandler.java [hunk 2]
```diff
@@ -46,39 +46,57 @@ public void startSource(ServerCallStreamObserver<GetEventStreamResponse> respons
...
+                            } catch (Exception e) {
+                                throw new RuntimeException(e);
+                            }
+                        }
                         // check whether the send queue has room for new messages
-                        while (!responseObserver.isReady() && !Context.current().isCancelled()) {
-                            // wait a bit to avoid OOM
-                            Thread.sleep(500);
+                        if (responseObserver.isReady()) {
+                            try {
```
Comment (StrikeW) [inline java/connector-node/risingwave-connector-service/src/main/java/com/risingwave/sourcenode/core/DbzSourceHandler.java:66]: No need to use two try-catch block, we can unify them to a single one.

File: java/connector-node/risingwave-connector-service/src/main/java/com/risingwave/sourcenode/core/DbzSourceHandler.java [hunk 3]
```diff
@@ -46,39 +46,57 @@ public void startSource(ServerCallStreamObserver<GetEventStreamResponse> respons
...
+                                            String.valueOf(config.getSourceId()),
+                                            resp.getEventsCount());
+                                    LOG.debug(
+                                            "Engine#{}: emit one chunk {} events to network ",
+                                            config.getSourceId(),
+                                            resp.getEventsCount());
+                                    responseObserver.onNext(resp);
+                                }
+                            } catch (InterruptedException e) {
+                                throw new RuntimeException(e);
```
Comment (StrikeW) [inline java/connector-node/risingwave-connector-service/src/main/java/com/risingwave/sourcenode/core/DbzSourceHandler.java:84]: Why remove the previous error log?

Comment (WillyKidd) [reply java/connector-node/risingwave-connector-service/src/main/java/com/risingwave/sourcenode/core/DbzSourceHandler.java]: mistake...

**Review by StrikeW — APPROVED** [review]
LGTM

**Pull request merged**